### ISSUE TITLE
Catalog Compilation API invocation rejecting hiera config(?)

### ISSUE THREAD

**Issue by mschuchard** [github/octocatalog-diff#106]
I hit this a while ago and took a long break to work on a bunch of other projects. I came back to this and it still seems really weird to me.

I am executing an RSpec test against an API call against octocatalog-diff. The method invocation is:

```ruby
OctocatalogDiff::API::V1.catalog(options)
```

where options is assigned by the following config file:

```ruby
# This is a configuration file for octocatalog-diff (https://github.com/github/octocatalog-diff).
module OctocatalogDiff
  # Configuration class.
  class Config
    def self.config
      settings = {}
      fixtures_dir = File.dirname(__FILE__) + '/'

      settings[:hiera_config] = fixtures_dir + 'hiera.yaml'
      settings[:hiera_path] = fixtures_dir + 'hieradata'
      settings[:fact_file] = fixtures_dir + 'hieradata/good.yaml'
      settings[:puppet_binary] = '/usr/local/bin/puppet'
      settings[:bootstrapped_to_dir] = fixtures_dir

      settings
    end
  end
end
```

and then a `:node` is appended via:

```ruby
options[:node] = node
```

where `node` is `good.example.com`.

The hiera config is:

```yaml
---
:backends: yaml
:yaml:
  :datadir: "hieradata"
:hierarchy:
  - "common"
```

The `hieradata` contains no `common.yaml`. The `manifests/site.pp` is:

```puppet
node good.example.com {
  notify { 'hello world': }
}

node syntax_error.example.com {
  notify { 'goodbye world' }
}
```

Given all of this, the following error is thrown:

> Failure/Error: OctocatalogDiff::API::V1.catalog(options)
     OctocatalogDiff::Errors::CatalogError:
       Catalog for 'to' (-) failed to compile with NoMethodError: undefined method `each' for "yaml":String
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/util/catalogs.rb:209:in `add_parallel_result'
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/util/catalogs.rb:112:in `block in build_catalog_parallelizer'
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/util/catalogs.rb:106:in `each'
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/util/catalogs.rb:106:in `build_catalog_parallelizer'
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/util/catalogs.rb:29:in `catalogs'
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/api/v1/catalog-compile.rb:35:in `catalog'
     # /var/lib/gems/2.3.0/gems/octocatalog-diff-1.0.4/lib/octocatalog-diff/api/v1.rb:15:in `catalog'
     # ./lib/puppet-check/regression_check.rb:12:in `block in smoke'
     # ./lib/puppet-check/regression_check.rb:10:in `each'
     # ./lib/puppet-check/regression_check.rb:10:in `smoke'
     # ./spec/puppet-check/regression_check_spec.rb:23:in `block (3 levels) in <top (required)>'

1. It seems to be grumpy about the hiera config file based on my debugging up to this point, but that minimal functioning example seems completely valid, especially since it is pared down from the default for Hiera 3. Is octocatalog-diff expecting a 4 or 5 convention hiera config?
2. The error seems to be thrown at https://github.com/github/octocatalog-diff/blob/384085155a5324384638ceb6737b24698e758ba2/lib/octocatalog-diff/util/catalogs.rb#L207-L209. Since there are odd values for `key` and `branch` here, is the implication that this is not going to work unless the bootstrap dir is specifically a composite git repo?

**Comment by kpaulisse** [comment]
It's expecting `backends` in hiera.yaml to be an array and not a string.

When I updated that, it appears to be right:

```
Kevins-MacBook-Pro-3:octocatalog-diff kpaulisse$ OCTOCATALOG_DIFF_CONFIG_FILE=/tmp/ms/ocd.cfg ./bin/octocatalog-diff --basedir /tmp/ms -n foo --no-parallel --catalog-only -d
D, [2017-05-03T00:24:09.896855 #46490] DEBUG -- : Running octocatalog-diff 1.0.4 with ruby 2.3.1
D, [2017-05-03T00:24:09.896934 #46490] DEBUG -- : Command line arguments: ["--basedir", "/tmp/ms", "-n", "foo", "--no-parallel", "--catalog-only", "-d"]
D, [2017-05-03T00:24:09.896978 #46490] DEBUG -- : Running on host Kevins-MacBook-Pro-3.local (x86_64-darwin15)
D, [2017-05-03T00:24:09.897037 #46490] DEBUG -- : Compiling catalog for foo
D, [2017-05-03T00:24:09.897446 #46490] DEBUG -- : Initialized OctocatalogDiff::Catalog::Noop for from-catalog
D, [2017-05-03T00:24:09.897485 #46490] DEBUG -- : Initialized OctocatalogDiff::Catalog::Computed for to-catalog
D, [2017-05-03T00:24:09.897534 #46490] DEBUG -- : Begin build_catalog for origin/master
D, [2017-05-03T00:24:09.897573 #46490] DEBUG -- : Setting up Puppet catalog build for origin/master
D, [2017-05-03T00:24:09.897607 #46490] DEBUG -- : Catalog for origin/master will be built with OctocatalogDiff::Catalog::Noop
D, [2017-05-03T00:24:09.897663 #46490] DEBUG -- : Catalog for origin/master successfully built with OctocatalogDiff::Catalog::Noop in 2.0e-06 seconds
D, [2017-05-03T00:24:09.897715 #46490] DEBUG -- : Success build_catalog for origin/master
D, [2017-05-03T00:24:09.897743 #46490] DEBUG -- : Begin build_catalog for .
D, [2017-05-03T00:24:09.897755 #46490] DEBUG -- : Setting up Puppet catalog build for .
D, [2017-05-03T00:24:09.897767 #46490] DEBUG -- : Catalog for . will be built with OctocatalogDiff::Catalog::Computed
D, [2017-05-03T00:24:09.897788 #46490] DEBUG -- : Calling build for object OctocatalogDiff::Catalog::Computed
D, [2017-05-03T00:24:09.897849 #46490] DEBUG -- : Start retrieving facts for foo from OctocatalogDiff::Catalog::Computed
D, [2017-05-03T00:24:09.897901 #46490] DEBUG -- : Retrieving facts from /tmp/ms/hieradata/good.yaml
D, [2017-05-03T00:24:09.899136 #46490] DEBUG -- : Success retrieving facts for foo from OctocatalogDiff::Catalog::Computed
D, [2017-05-03T00:24:09.899889 #46490] DEBUG -- : Symlinked /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production -> /tmp/ms/
W, [2017-05-03T00:24:09.900332 #46490]  WARN -- : WARNING: Hiera datadir for yaml doesn't seem to exist at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/tmp/ms/hieradata
D, [2017-05-03T00:24:09.901166 #46490] DEBUG -- : Installed hiera.yaml from /tmp/ms/hiera.yaml to /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/hiera.yaml
D, [2017-05-03T00:24:09.902208 #46490] DEBUG -- : Installed fact file at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var/yaml/facts/foo.yaml
D, [2017-05-03T00:24:10.796808 #46490] DEBUG -- : (to) Try 1 executing Puppet 3.8.7: /Users/kpaulisse/github/puppet/script/puppet3 master --compile foo --no-storeconfigs --factpath=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var/yaml/facts --facts_terminus=yaml --no-daemonize --no-ca --color=false --config_version="/bin/echo catalogscript" --environment=production --hiera_config=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/hiera.yaml --environmentpath=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments --vardir=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var --logdir=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var --ssldir=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var/ssl --confdir=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3
D, [2017-05-03T00:24:10.797594 #46490] DEBUG -- : ["Execute: /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-1dbaca8/puppet.sh master --compile foo --no-storeconfigs --factpath\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var/yaml/facts --facts_terminus\\=yaml --no-daemonize --no-ca --color\\=false --config_version\\=\\\"/bin/echo catalogscript\\\" --environment\\=production --hiera_config\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/hiera.yaml --environmentpath\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments --vardir\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var --logdir\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var --ssldir\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/var/ssl --confdir\\=/var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3"]
D, [2017-05-03T00:24:12.668672 #46490] DEBUG -- : ["STDERR: Error: Could not parse for environment production: Syntax error at '.' at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/manifests/site.pp:1 on node foo"]
D, [2017-05-03T00:24:12.668740 #46490] DEBUG -- : ["STDERR: Error: Could not parse for environment production: Syntax error at '.' at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/manifests/site.pp:1 on node foo"]
D, [2017-05-03T00:24:12.668761 #46490] DEBUG -- : ["STDERR: Error: Failed to compile catalog for node foo: Could not parse for environment production: Syntax error at '.' at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/manifests/site.pp:1 on node foo"]
D, [2017-05-03T00:24:12.668780 #46490] DEBUG -- : ["Exit status: 30"]
W, [2017-05-03T00:24:12.668892 #46490]  WARN -- : Puppet command failed: STDOUT:
STDERR:
  Error: Could not parse for environment production: Syntax error at '.' at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/manifests/site.pp:1 on node foo
  Error: Could not parse for environment production: Syntax error at '.' at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/manifests/site.pp:1 on node foo
  Error: Failed to compile catalog for node foo: Could not parse for environment production: Syntax error at '.' at /var/folders/dw/5ftmkqk972j_kw2fdjyzdqdw0000gn/T/d20170503-46490-14sslw3/environments/production/manifests/site.pp:1 on node foo
D, [2017-05-03T00:24:12.668945 #46490] DEBUG -- : (to) Catalog failed on try 1 in 2.766646 seconds
D, [2017-05-03T00:24:12.669016 #46490] DEBUG -- : Catalog for . failed with OctocatalogDiff::Catalog::Computed in 2.771209 seconds retries = 0
W, [2017-05-03T00:24:12.669051 #46490]  WARN -- : Failed build_catalog for .
/Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/util/catalogs.rb:202:in `add_parallel_result': Catalog for . failed to compile due to errors: (OctocatalogDiff::Errors::CatalogError)
[Puppet Error] Could not parse for environment production: Syntax error at '.' at manifests/site.pp:1 on node foo
[Puppet Error] Could not parse for environment production: Syntax error at '.' at manifests/site.pp:1 on node foo
[Puppet Error] Failed to compile catalog for node foo: Could not parse for environment production: Syntax error at '.' at manifests/site.pp:1 on node foo
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/util/catalogs.rb:112:in `block in build_catalog_parallelizer'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/util/catalogs.rb:106:in `each'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/util/catalogs.rb:106:in `build_catalog_parallelizer'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/util/catalogs.rb:29:in `catalogs'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/api/v1/catalog-compile.rb:35:in `catalog'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/api/v1.rb:15:in `catalog'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/cli.rb:170:in `catalog_only'
	from /Users/kpaulisse/github/octocatalog-diff/lib/octocatalog-diff/cli.rb:94:in `cli'
	from ./bin/octocatalog-diff:34:in `<main>'
```

**Comment by kpaulisse** [comment]
Looks like we need to add the possibility for this to be a string.

Please just appease it with an array for now, and I've filed an issue here: https://github.com/github/octocatalog-diff/issues/107

**Comment by mschuchard** [comment]
Awesome; thanks for the quick response.

I had to do a double-check myself since the key symbol is `:backends:` and not `:backend:` so now I was doubting myself some. Found the doc for it for Hiera 3 to confirm: https://docs.puppet.com/hiera/3.3/configuring.html#backends.

For some reason I thought octocatalog-diff would be leveraging Hiera itself for the config file parsing, but now I realize that doesn't make any sense from a software design perspective.

Closing this in favor of the new issue.
